website/project/model.py

```diff
@@ -2788,6 +2811,126 @@ def __repr__(self):
         return '<WatchConfig(node="{self.node}")>'.format(self=self)
 
 
+class MailingList(StoredObject):
+    """Discussions mailing list object for projects and components"""
+
+    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
+    changed = fields.BooleanField(default=False)
+
+    node_id = fields.StringField()
+    node_title = fields.StringField()
+    node_deleted = fields.BooleanField(default=False)
+
+    is_enabled = fields.BooleanField(default=False)
+
+    mailing_info = fields.DictionaryField()
+
+    @classmethod
+    def new_saved(cls, **kwargs):
+        """Creates a new mailing list and saves it so that it can be an attribute of a node"""
+        new_list = cls(**kwargs)
+        new_list.save()
+
+        return new_list
+
+    def set_node(self, node, save=True):
+        """Sets object variables to match node and allows it to start updating Mailgun"""
+        self.node_id = node._id
+        self.node_title = node.title
+        self.match_members(node.contributors, save=False)
+
+        try:
+            self.is_enabled = not node.parent
+        except AttributeError:
+            pass
+
+        if save:
+            self.save()
+
+    def match_members(self, node_contributors, save=True):
+        """ Matches list members with parent node
+        :param node_contributors: List of contributors on the node
+        :param save: Automatically saves if True
+        """
+        new_info = {}
+        for user in node_contributors:
+            new_info[user._id] = self.mailing_info.get(user._id) or {
+                'address': user.email,
+                'subscribed': True
+            }
+        self.mailing_info = new_info
+
+        if save:
+            self.save()
+
+    def add_member(self, user):
+        if user.is_registered:
+            self.mailing_info[user._id] = {
+                'address': user.email,
+                'subscribed': True,
+            }
+
+    def remove_member(self, user):
+        del self.mailing_info[user._id]
+
+    @property
+    def emails(self):
+        return set([info['address'] for info in self.mailing_info.values()])
+
+    @property
+    def subscriptions(self):
+        return set([info['address'] for info in self.mailing_info.values() if info['subscribed']])
+
+    def enable(self, save=False):
+        if self.node_deleted:
+            raise NodeStateError('Deleted nodes cannot have active mailing lists')
+
+        self.is_enabled = True
+
+        if save:
+            self.save()
+
+    def disable(self, deleted=False, save=False):
```

Comment:
Don't use the `save` flag on objects unless there is specific reason to do so, or if it's used elsewhere on the same object.

The only place I'm *sure* it's used now is on `Node`.